cpp/oneapi/dal/backend/primitives/objective_function/logloss_functors.hpp

```diff
@@ -0,0 +1,122 @@
+
+using comm_t = backend::communicator<spmd::device_memory_access::usm>;
+
+template <typename Float>
+class logloss_hessian_product : public base_matrix_operator<Float> {
+    friend dal::detail::pimpl_accessor;
+
+public:
+    logloss_hessian_product(sycl::queue& q,
+                            const table& data,
+                            Float L2 = Float(0),
+                            bool fit_intercept = true,
+                            std::int64_t bsz = -1);
+    logloss_hessian_product(sycl::queue& q,
+                            comm_t comm,
+                            const table& data,
+                            Float L2 = Float(0),
+                            bool fit_intercept = true,
+                            std::int64_t bsz = -1);
+    sycl::event operator()(const ndview<Float, 1>& vec,
+                           ndview<Float, 1>& out,
+                           const event_vector& deps) final;
+    ndview<Float, 1>& get_raw_hessian();
+
+private:
+    void reserve_memory();
+
+    sycl::event compute_with_fit_intercept(const ndview<Float, 1>& vec,
+                                           ndview<Float, 1>& out,
+                                           const event_vector& deps);
+    sycl::event compute_without_fit_intercept(const ndview<Float, 1>& vec,
+                                              ndview<Float, 1>& out,
+                                              const event_vector& deps);
+    sycl::queue& q_;
+
+    comm_t comm_;
+    const table data_;
+    Float L2_;
+    bool fit_intercept_;
+    ndarray<Float, 1> raw_hessian_;
+    ndarray<Float, 1> buffer_;
+    ndarray<Float, 1> tmp_gpu_;
+    const std::int64_t n_;
+    const std::int64_t p_;
+    const std::int64_t bsz_;
+    dal::detail::pimpl<sparse_matrix_handle> sp_handle_;
+};
+
+template <typename Float>
+class logloss_function : public base_function<Float> {
+    friend dal::detail::pimpl_accessor;
+
+public:
+    logloss_function(sycl::queue& queue,
+                     const table& data,
+                     const ndview<std::int32_t, 1>& labels,
+                     Float L2 = 0.0,
+                     bool fit_intercept = true,
+                     std::int64_t bsz = -1);
+    logloss_function(sycl::queue& queue,
+                     comm_t comm,
+                     const table& data,
+                     const ndview<std::int32_t, 1>& labels,
+                     Float L2 = 0.0,
+                     bool fit_intercept = true,
+                     std::int64_t bsz = -1);
+    Float get_value() final;
+    ndview<Float, 1>& get_gradient() final;
+    base_matrix_operator<Float>& get_hessian_product() final;
+
+    event_vector update_x(const ndview<Float, 1>& x,
+                          bool need_hessp = false,
+                          const event_vector& deps = {}) final;
+
+private:
+    void reserve_memory();
+
+    sycl::queue& q_;
+    comm_t comm_;
+    const table data_;
```

Comment:
probably, it will better if you order it by some logic, like
comm
q
tables,
ndarrays
Float values etc